### ISSUE TITLE
Binding model in WGSL does not respect encapsulation

### ISSUE THREAD

**Issue by litherum** [gpuweb/gpuweb#2426]
Suppose you had two different compute shaders in the same file. It would be pretty natural to want to write something like

```
/* This is my FOO shader */
[[group(0), binding(0)]] var<storage> buffer0  : BufferInput;
[[stage(compute)]] fn foo() {
    ... use buffer0 ...
}

/* This is my BAR shader */
[[group(0), binding(0)]] var<storage> buffer0  : BufferInput;
[[group(0), binding(1)]] var texture0 : texture_2d<f32>;
[[stage(compute)]] fn bar() {
    ... use buffer0 and texture0 ...
}
```

`foo()` and `bar()` are unrelated, yet his will fail to compile, because there are two variables in the same scope with the same name (`buffer0`). The fact that you have to either have the two bindings have different variable names, or you have to deduplicate them, seems like a violation of encapsulation. It would also be unfortunate for an author to split `foo()` and `bar()` into two different calls to `createShaderModule()` because then any shared helper functions would have to be duplicated and compiled twice.

**Comment by kvark** [comment]
We can only do so much for catering the use case of people mixing text together. My vision is that users who do want to concatenate shaders should take extra care about the dependencies. If A needs B, then B has to be inserted before A. If A and B both need C (like in this case), then C should be a separate piece of text that is inserted before A and B.

TL;DR: I don't think there is anything we should be doing for this issue. Allowing name collisions seems very troublesome.
Perhaps, there could be syntax for scoping global things?
```rust
{
  /* This is my FOO shader */
  [[group(0), binding(0)]] var<storage> buffer0 : BufferInput;
  [[stage(compute)]] fn foo() {
      ... use buffer0 ...
}
{
  /* This is my BAR shader */
  [[group(0), binding(0)]] var<storage> buffer0 : BufferInput;
  [[group(0), binding(1)]] var texture0 : texture_2d<f32>;
  [[stage(compute)]] fn bar() {
      ... use buffer0 and texture0 ...
  }
}
```

**Comment by jdashg** [comment]
Yeah, scoping is the normal solution, and expanding lexical scoping to the top level seems like a good direction.

**Comment by magcius** [comment]
Adding scoping in this manner will have knock-on effects. Let's consider all the ramifications before adding something like this. I have my doubts about this compilation model being that helpful in practice.

**Comment by litherum** [comment]
Right - the reason I didn't immediately propose scoping at the top level is it doesn't work with shared helper functions.

But maybe that's okay?

I don't know; I honestly don't know what a good solution to this problem would be.

**Comment by Kangz** [comment]
Would support for namespaces help with this kind of issue? It could be a post-V1 feature.

**Comment by litherum** [comment]
Upon reflection, there might actually be value in adding a namespace feature now, because of both this issue and https://github.com/gpuweb/gpuweb/issues/600 are in the V1 milestone.

The main motivation is a cost/benefit tradeoff:
- WGSL already has order independence at the top level, so implementations must already retain a data structure that represents the whole program.
- WGSL already has lexical scoping, so implementations must already be able to handle scope chains and name shadowing
- Namespaces would allow for a natural solution for both this issue and https://github.com/gpuweb/gpuweb/issues/600

So, the benefit is considerable, and the cost is probably pretty small, since almost all the relevant infrastructure for supporting namespaces already exists.

So here's a super simple proposal, taken pretty much wholesale from C++, described via examples:
```
namespace Foo {
    fn bar() {
        ...
    }

    @stage(compute) fn baz() {
        bar();
    }
}

@stage(compute) fn baz() {
    Foo::bar();
}
```
A pipeline would hook up the inner `baz()` as its entry point by passing in the string `Foo::baz` to `create*Pipeline()`.

Presumably, we'd also put the standard library in a namespace:
```
fn foo() {
    wgsl::smoothstep(...);
}
```

In order to not keep having to repeat `wgsl::` (just like you don't have to repeat a prefix today) we'd probably want a `using` declaration. This would essentially create aliases for everything in the named namespace and put them in the current namespace:
```
using namespace wgsl;

fn foo() {
    smoothstep(...);
}
```

We'd maintain forward compatibility by putting new standard library additions into their own namespace:
```
fn foo() {
    wgsl::smoothstep(...);
    wgsl2::raytrace(...);
}
```

It would work alongside the `enable` directive:
```
enable f16;

fn foo() {
    var myHalf: f16 = ...;
    wgsl::smoothstep(... myHalf ...);
}
```

This encapsulation issue would be solved as such:
```
namespace FOO {
    @group(0), @binding(0) var<storage> buffer0 : BufferInput;
    @stage(compute) fn foo() {
        ... use buffer0 ...
    }
}

namespace BAR {
    @group(0), binding(0) var<storage> buffer0 : BufferInput;
    @group(0), binding(1) var texture0 : texture_2d<f32>;
    @stage(compute) fn bar() {
        ... use buffer0 and texture0 ...
    }
}

fn helper() {
    ... use BAR::buffer0 ...
}
```

In conclusion, this kind of thing seems like a natural solution to 2 filed issues, is something that authors will be familiar with, and doesn't seem very difficult to implement. It seems like a good fit.

**Comment by ben-clayton** [comment]
Seems reasonable. My initial thoughts / questions:
* May I suggest `.` for accessing inside a namespace instead of `::`? `::` is very C++, and `.` is more common considering the number of other languages that have influenced WGSL.
* Can you use `using` with a user-declared namespace?
* What happens when there's a symbol collision?
* Do we want to support nested namespaces?
   `namespace A.B {` 
   or 
   `namespace A { namespace B {`
* Do we want to support 'partial' namespaces? Might be handy for the string-concatenation case. e.g.:
   ```
   namespace A { type stuff = i32; } 
   ...
   namespace A { type moar_stuff = i32; }
   ```
   If so, do we want to prevent users adding to the standard namespace(s)?
* Do we move all the existing builtins into a `wgsl` namespace for V1?

**Comment by litherum** [comment]
So many questions! I'll try to make a first stab at answers:

1. Sure, sounds good.
2. I generally dislike anything that makes built-in facilities special, so I'd propose "yes."
3. I don't think I have opinions; whatever the group prefers is fine.
4. Compiler error
5. I don't think I have opinions; whatever the group prefers is fine.
6. I don't think I have opinions; whatever the group prefers is fine.
7. I think that would be a good idea - if we assume that additions would be in their own namespace, then it would be balanced if the existing ones were moved too.

**Comment by litherum** [comment]
I guess there's a related question: How does order independence at the top level work?

I think my proposed solution would be "contents of namespaces would be considered 'at the top level' for the purposes of order independence"

**Comment by kainino0x** [comment]
I argue:
- WGSL should be a living spec just like the rest of the web platform. The set of available language features is tied strictly to the browser version.
- There should be a way to detect language features from JS, because checking for errors on createShaderModule is hard. Eventually people will stop checking these once they're available on all their target browsers on <https://caniuse.com>.
- Namespaces for the standard library (prelude): very tempting but not actually helpful
- (Namespaces for hardware-dependent features: no real opinion but don't think it's necessary)
- Prelude namespaces don't solve the identifier collision problem; prelude namespace identifiers can just as easily collide with user identifiers
- To be non-breaking, new additions to the prelude should be shadowable by user code. (Maybe the whole prelude should be shadowable, maybe not.)

I propose:
- No named namespaces (user-defined or prelude) for now: avoids `.`/`::` syntax, no need for `using`.
- Only have anonymous namespaces - item-level "scopes" - works just like statement-level scopes (which can't be named or reached into).
- Nested scopes: yes.
- Entry point names still have to be globally unique.
- syntax either `namespace { }` or `{ }`. Former is nice because it is more explicit and gives us a syntax for named namespaces. Latter is nice because it's the same as statement-level scopes.

```
namespace {
    @group(0), @binding(0) var<storage> buffer0 : BufferInput;
    @stage(compute) fn foo() {
        ... use buffer0 ...
        ... call helper() ...
    }
}

namespace {
    @group(0), binding(0) var<storage> buffer0 : BufferInput;
    @group(0), binding(1) var texture0 : texture_2d<f32>;
    @stage(compute) fn bar() {
        ... use buffer0 and texture0 ...
        ... call helper() ...
    }

    fn helper() {
        ... use buffer0 ...
    }
}
```

**Comment by magcius** [comment]
Namespaces seem like a weird feature to check for. What's the recommended strategy? If a user tests for namespaces, and they're not supported, their options are: 1. rewrite their shaders to not use namespaces, in which case they won't need the feature to begin with, 2. run some compiler tool which will do that for you, in which case, you also don't need the feature to begin with, 3. show an error message for an unsupported browser.

**Comment by litherum** [comment]
I think I'd be comfortable with @kainino0x's proposal.

@magcius I think we're discussing how to put namespaces into V1, rather than postponing it. This issue is in the V1 milestone. So I think we're all in agreement?

**Comment by kainino0x** [comment]
I just realized that the first comment in the thread suggested lexical scoping at the top level already! so first credit to @kvark :)

> Namespaces seem like a weird feature to check for. What's the recommended strategy? If a user tests for namespaces, and they're not supported, their options are: 1. rewrite their shaders to not use namespaces, in which case they won't need the feature to begin with, 2. run some compiler tool which will do that for you, in which case, you also don't need the feature to begin with, 3. show an error message for an unsupported browser.
> 
> None of these seem like good options?

The stuff I wrote about detecting features wasn't about namespaces as a feature, but instead trying to prove that namespaces don't solve anything for the prelude, even for prelude functions we add post-v1.

For most core language features added post-v1 though, I absolutely agree with your analysis. This would be just for your case 3. We would add some way to feature-detect from JS just so that applications that assume a feature could skip their webgpu backend on the long tail of browsers where the feature is not supported - feature detection in this way is the standard alternative to user agent detection.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-vf2f-2022-02-0809-minutes#mm-how-to-encapsulateassociate-resources-with-a-module-without-them-colliding-in-name-eg-two-modules-with-different-resources-but-using-the-same-name-like-u_tex-httpsgithubcomgpuwebgpuwebissues2426">WGSL meeting minutes 2022-02-08</a></summary>



* (DN: #600 was the issue I filed on this, which we resolved to have features, and maybe namespaces but we never went back to that.)
* [https://github.com/gpuweb/gpuweb/issues/2426#issuecomment-1032109816](https://github.com/gpuweb/gpuweb/issues/2426#issuecomment-1032109816) 
* MM: Posted last night, so it’s a little soon.
* MM: With namespaces, we can put new stdlib things in namespaces, and new ones won’t conflict then. We might want to add a new syntax e.g. array types, and namespaces don’t solve that, but it helps with new builtins.
* MM: They can also be useful for scoping global variables between entrypoints, too.
* MM: I think this work in conjunction with the enable directive. So if we add e.g. array brackets, we still use the enable directive.
* MM: To JB’s question, new features would come in the form of new namespaces/additions and new directives.
* DM: I remember there was a discussion about namespaces a while ago, and we decided that we can just put builtin in root, so can someone explain the gap there?
* MM: I think we couldn’t agree on how to divvy up namespaces, so we compromised on putting them in root. I do think that we have new problems to solve here, so namespaces have more value this time/for this.
* JB: Background: every other language design I’ve been involved with, it’s been hugely fraught.  We should not beat ourselves up if we struggle with.  Substantively, feel it’s easier for me to keep track of things when theres a linear progression of additions to the language. There’s a lot of ergonomic benefit to having versions over having a series of options that can be enabled individually.
* BC: Separate namespace per extension: Not sure you need it.  If part of the set of rules of namespaces that there are reserved namespaces, then you can ensure there are no collisions, as long as extension authors don’t collide.
* BC: Am vague about this proposal an “requires”. Long time ago, we said we’d have a block of { .. } with requires, and you could put a feature in that program.  My argument for that was linking needing to link across those boundaries. Not clear where the proposal stands on that. So really want the requires/enables at the top of the source.
* BC: With Jim on the unpleasantness of having a bunch  of independent named features. That’s complex for implementers and users. But versioning doesn’t play nice with features that only exist in certain implementations. E.g. f16 we really like, but we can’t make it core overnight.  Have to wait for the broadbase of hardware to catch up to features that people wanted yesterday.
* AB: Still see namespaces as purely additive.  We can release the lang without ns.  Future, add features with an enable.  Once you add an enable, it’s ok if your code got broke because you opted in.  Feature soup seems inevitable. It’s hard to get core versions out if the only things you could add have to be supported everywhere. It would move extremely slowly.
* MM: About linear progression.  Think that’s reasonable.  For features like f16, probably doesn’t belong in linear progression of the language.  E.g. raytracing doesn’t belong.  Something like unicode identifiers *does* belong in next version of the language.  Think that idea naturally fits here.  E.g. there’s an extension named “wgsl2”, then “wgsl3” that each is superset of the previous. Not proposing a separate namespace per extension.  If adding new builtins to standard library, could have new enable, without new namespace.  Keep adding them to the existing namespace.
* MM: Not trying to change the “enable” being at the top.  This is just about name resolution.
* MM: An “enable” is conceptually a namespace. You import new functionality into the global namespace.  One of the reason namespaces is good is it gives shader authors the ability to do something the standard library  / implementation already has.  It also supports feature of grouping things; that’s a shader author problem that is not a problem for the standard problem.
* KN: Namespaces good.  Jim’s right “modules” system is the hard part.
* KN: Thought we were happy to put all standard library stuff into the global namespace.  Including extra helpers, for hardware independent features.  Don’t think it’s a breaking change.  Think we should continue putting hw-independent things in the global namespace.
* KN: Try to stay away from language versions. That is the “way of the web”.  E.g. unicode identifiers. Don’t think you need an enable.  Non-breaking language change should just roll out as browsers implement them. Don’t think a browser should have to validate a rule like that.
* MM: WGSL is fundamentally less dynamic than JavaScript.  In JS you write an if statement to see if you can use a feature. Can’t do that in WGSL.  Example: unicode identifiers. Authors will want to detect whether they can use that feature.  To detect whether unicode identifiers can be use, the author will try to compile a shader with a unicode identifier.
* KN: Put the feature reporting/detection outside the shader. You would be able to detect availability of Features via JS, but it’s unconditionally enabled in shaders, and just works.
* MM: So the information flows from the browser to the author.  (Not the author providing the info to the implementation)
* MM: I worry about fingerprinting, but that sounds reasonable.
* KN: There is no fingerprinting beyond the browser version. It’s strictly tied to the browser version and there is no further identifying info.
* KG: Regarding versioning, I think there is a middle ground. Might choose to bundle a bunch of extensions into a core version of the language.  Could be a very far future thing; can tackle it when we get to it.  Think the enable directive is sufficient for now.  Can decide versioning model later. Allow the luxury of not solving that problem now.  Example: ES 2.0 vs. ES 3.0 is really valuable, but the gap is huge. (Large granularity)
* BC: To Kai’s idea of a rolling language feature set.  Two things seemed in conflict: don’t need a namespace for library. How do you deal with symbol collisions, e.g. user code declares something that is later added to the core.  If you don’t declare the version/features of the code in the shader code itself, then that makes offline tooling difficult.
* KN: For second thing, that’s a fair point.  But not sure why tooling would need to do anything but the latest. 
* KN: For collisions, we can find another solution to that. We have to add namespaces too, that will add an identifier. Don’t think it’s a different problem. Can say for things added in the future, they can be shadowed by things the user defines. For example. 
* DM: Concern about that approach of  “your shader will fail anyway”. As a developer, I may be aware of 3 language features, and have shaders with 3 variants.  Bunch of work to try them all against all. Seems unfortunate.  Tragedy is it’s asynchronous progression.  Might help on the API side to have a true/false answer for whether a feature exists.
* JB: Specific example of unicode identifiers irks me. Nobody is going to keep two copies of their shaders.  Will just keep one copy around, the base case.
* MM: Think authors will have two versions: an old vs. a new (that uses many new features together).
* KN: People will use caniuse.com to see what has broad coverage among browsers.
* MM: Unicode identifiers is too small a granularity feature. Would want to bundle that with a bunch of other things that are independent of hardware features.
* KN:...
* MM: I can buy that.
* MM: Problem I’m trying to solve is a zoo of small features.  Rolling works when many browsers implement things in roughly same order
* KG: There is a long tail. We have a product “ESR” with support for year-old release. We have to care some. Would like to avoid situation where people don’t have a good idea about when something is going to work. Unsatisfactory is quickly checking can-i-use, try it once, the ship it.  Nice thing in JS small code snippet to check the feature.
* KN: The more we can make our problem look like the web platform’s problem, the better.
* MM: As long as the feature is feature-detectable.  Compile a shader and see if it fails is reasonable as long as the error/pass cases are consistently checkable.
* KN: think prefer the browser just reports which features are available, no async compile tests.
* KG: I feel that user namespaces have value for e.g. the encapsulation problem for conflicting var in entrypoints. I don’t know as we want to sign up more generally to agree to use namespaces for spec additions. I would like to see a concrete example, where namespaces solves a problem there, rather than saying “generally we will solve it with namespaces”. But, if we want to use namespaces *for* versioning, we do then need them in the initial release.
* MM: 
* JB: Don’t be guided by implementation complexity.  Casual design decisions by language designers have long term pain for users to deal with.
* KN: Do we need to have named namespaces for the encapsulation problem? Anon namespaces would solve that with lower complexity/impact.
* MM: I don’t necessarily think that namespaces are directly related to modules per se. I think that namespaces are simpler than that.
* DN: For visibility, everything is currently visible, so that’s sort of a decision about how to do modules. Second, we do need a solution for how to reference in-namespace entrypoints and vars from the API side.
* MM: Can use the fully-qualified names.
* DN: Ok, though anon spaces won’t work for conflicts there.
* KN: Was imagining that entry point names would still have to be unique across the shader. Encapsulation doesn’t seem as important for entry point names (vs bindings).
* MM: Not a huge fan, but acceptable.
* MM: If we do make them unnamed, we probably shouldn’t call this NAMEspaces
* DN: “extra scopes”?


</details>

**Comment by dneto0** [comment]
Is this critical for v1?

I've never seen HLSL code use namespaces.  I didn't even know HLSL had namespaces until I dug quite far.
I didn't see usable documentation about it on the Microsoft site [here](https://docs.microsoft.com/en-us/windows/win32/direct3dhlsl/dx-graphics-hlsl-language-syntax).
* DXC supports namespaces.
* Glslang has an HLSL compilation path, which is definitely in serious production use, but doesn't fully support namespaces. [reference](https://github.com/KhronosGroup/glslang/issues/362).  In fact, the test case for it in the Glslang suite was buggy.   [my fix](https://github.com/KhronosGroup/glslang/pull/2891)

**Comment by kainino0x** [comment]
The motivating use case is multiple entry points. Do you know how common those are in HLSL?

Of course, people can always work around not having namespaces, in HLSL or WGSL, so I wouldn't personally call it a hard V1 requirement.

**Comment by kvark** [comment]
I worked with HLSLs having dozens of entry points, but these entry points had a lot in common. They'd not use completely different bindings under the same name, for example. That's the model I imagine for WGSL as well: not that it's going to be a huge library of independent shaders, but that each WGSL module would have a bunch of related entry points.

**Comment by dneto0** [comment]
I still don't buy this is necessary for V1.

Very large C programs don't need namespacing and scoping like this. For example, the Linux kernel is a very large program.

Programmers will put things into the same translation unit when it makes sense to, based on increasing cohesion and decreasing coupling.

**Comment by ben-clayton** [comment]
> revert order-independent module declarations for V1, since I believe that order-independent module declarations provide little benefit without a scope/namespaces tool

We (Google) have found a number of use cases for out of order declarations, which are orthogonal to scopes & namespaces. As this feature is in active use, we'd strongly prefer to keep them.

**Comment by kdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/wgsl-2022-02-22-minutes#binding-model-in-wgsl-does-not-respect-encapsulation-2426-">WGSL meeting minutes 2022-02-22</a></summary>



* (v1?)
* MM: Situation I’m worried about is if it’s optional, but then developers don’t use it because it’s not incore.
* MM: Think the cost is low.  Important to encourage authors to put as much inside a single module as possible.
* DN: The optionality is an argument that this isn’t necessary for v1.
* MM: Minimal proposal is braces at top-level (unnamed namespaces). Introduces scoping.
* DN: What about things nameable by the API.
* MM: Kai’s answer: rule is that anything nameable by the API (entry point name, and overrides), are not allowed to collide.  So they’re like their at a top-level namespace. 
* AB: Does this solve people’s problem.  Are entry point collisions a problem.  This only solves the problem for helper functions, resource bindings, and module-scope var and let.  What is the extent that is being envisioned here that is being combined in one module. The only name collisions you’re saving here are independent bindings.
* MM: Named namespaces was the original idea.  Think they would be better.  Current idea is the direction from the issue discussion.  Current idea can be extended to named namespaces. 
* AB: Why is it more complicated to prefix just the entry point, and not everything.
* KG: Your concatenator is much simpler if it only has to rename the entry point and overrides.
* JB: Find Alan’s points persuasive. It’s only a partial solution. Think a real solution is the named namespaces. Think we should spec and land named namespaces first.
* KG: Let’s work on unnamed spaces.
* **MM to write PR**


</details>
